src/python/pants/core/goals/check.py

```diff
@@ -0,0 +1,230 @@
+
+
+@frozen_after_init
+@dataclass(unsafe_hash=True)
+class CheckResults:
+    """Zero or more CheckResult objects for a single type checker.
+
+    Typically, type checkers will return one result. If they no-oped, they will return zero results.
+    However, some type checkers may need to partition their input and thus may need to return
+    multiple results.
+    """
+
+    results: Tuple[CheckResult, ...]
+    typechecker_name: str
+
+    def __init__(self, results: Iterable[CheckResult], *, typechecker_name: str) -> None:
+        self.results = tuple(results)
+        self.typechecker_name = typechecker_name
+
+    @property
+    def skipped(self) -> bool:
+        return bool(self.results) is False
+
+    @memoized_property
+    def exit_code(self) -> int:
+        return next((result.exit_code for result in self.results if result.exit_code != 0), 0)
+
+
+class EnrichedCheckResults(CheckResults, EngineAwareReturnType):
+    """`CheckResults` that are enriched for the sake of logging results as they come in.
+
+    Plugin authors only need to return `CheckResults`, and a rule will upcast those into
+    `CheckResults`.
+    """
+
+    def level(self) -> Optional[LogLevel]:
+        if self.skipped:
+            return LogLevel.DEBUG
+        return LogLevel.ERROR if self.exit_code != 0 else LogLevel.INFO
+
+    def message(self) -> Optional[str]:
+        if self.skipped:
+            return f"{self.typechecker_name} skipped."
+        message = self.typechecker_name
+        message += (
+            " succeeded." if self.exit_code == 0 else f" failed (exit code {self.exit_code})."
+        )
+
+        def msg_for_result(result: CheckResult) -> str:
+            msg = ""
+            if result.stdout:
+                msg += f"\n{result.stdout}"
+            if result.stderr:
+                msg += f"\n{result.stderr}"
+            if msg:
+                msg = f"{msg.rstrip()}\n\n"
+            return msg
+
+        if len(self.results) == 1:
+            results_msg = msg_for_result(self.results[0])
+        else:
+            results_msg = "\n"
+            for i, result in enumerate(self.results):
+                msg = f"Partition #{i + 1}"
+                msg += (
+                    f" - {result.partition_description}:" if result.partition_description else ":"
+                )
+                msg += msg_for_result(result) or "\n\n"
+                results_msg += msg
+        message += results_msg
+        return message
+
+
+@union
+class CheckRequest(StyleRequest):
+    """A union for StyleRequests that should be type-checkable.
+
+    Subclass and install a member of this type to provide a linter.
+    """
+
+
+class CheckSubsystem(GoalSubsystem):
+    name = "check"
+    help = "Run type checking or the lightest variant of compilation available for a language."
+
+    deprecated_options_scope = "typecheck"
+    deprecated_options_scope_removal_version = "2.10.0.dev0"
```

Comment:
This can be 2.9.0.dev0 no?

Reply:
Yeah agreed. This deprecation doesn't seem painful enough to need 2 major releases.

Reply:
Yea, it could be. Because this changes something so fundamental, and it's cheap to keep as deprecated, I biased toward the longer deprecation. Fine either way though.

Reply:
I'm fine either way, your choice.